File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/list-ops/list-ops.js
//
// This is only a SKELETON file for the 'List Ops' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class List {
  constructor(elements = []) {
    this.elements = [...elements];
  }

  append(otherList) {
    return new List([...this.elements, ...otherList.elements]);
  }

  concat(...lists) {
    let newElements = [...this.elements];
    for (const list of lists) {
      newElements = [...newElements, ...list.elements];
    }
    return new List(newElements);
  }

  filter(predicate) {
    let filtered = [];
    for (const el of this.elements) {
      if (predicate(el)) {
        filtered = [...filtered, el];
      }
    }
    return new List(filtered);
  }

  map(fn) {
    let mapped = [];
    for (const el of this.elements) {
      mapped = [...mapped, fn(el)];
    }
    return new List(mapped);
  }

  length() {
    let count = 0;
    for (const _ of this.elements) {
      count++;
    }
    return count;
  }

  foldl(fn, acc) {
    let result = acc;
    for (const el of this.elements) {
      result = fn(result, el);
    }
    return result;
  }

  foldr(fn, acc) {
    let result = acc;
    for (let i = this.elements.length - 1; i >= 0; i--) {
      result = fn(result, this.elements[i]);
    }
    return result;
  }

  reverse() {
    let reversed = [];
    for (let i = this.elements.length - 1; i >= 0; i--) {
      reversed = [...reversed, this.elements[i]];
    }
    return new List(reversed);
  }
}

File: src/2025-03-16-12-54-49--a-deepclaude-full-architect-python-0316-2/python/exercises/practice/list-ops/list_ops.py
def append(list1, list2):
    result = list1.copy()
    for item in list2:
        result.append(item)
    return result


def concat(lists):
    result = []
    for list in lists:
        for item in list:
            result.append(item)
    return result


def filter(function, list):
    result = []
    for item in list:
        if function(item):
            result.append(item)
    return result


def length(list):
    count = 0
    for _ in list:
        count += 1
    return count


def map(function, list):
    result = []
    for item in list:
        result.append(function(item))
    return result


def foldl(function, list, initial):
    result = initial
    for item in list:
        result = function(result, item)
    return result


def foldr(function, list, initial):
    if length(list) == 0:
        return initial
    return function(list[0], foldr(function, list[1:], initial))


def reverse(list):
    result = []
    for i in range(length(list) - 1, -1, -1):
        result.append(list[i])
    return result
